src/Tools/Source/CompilerGeneratorTools/Source/CSharpSyntaxGenerator/SoftReference`1.cs

```diff
@@ -0,0 +1,68 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System;
+using System.Diagnostics.CodeAnalysis;
+using System.Threading;
+
+namespace CSharpSyntaxGenerator
+{
+    /// <summary>
+    /// This class behaves like <see cref="WeakReference{T}"/>, with the added constraint that the referenced object
+    /// will not be reclaimed prior to the next Generation 2 garbage collection operation.
+    /// </summary>
+    /// <remarks>
+    /// The Generation 2 constraint only applies to the current target of the soft reference. If the target is replaced
+    /// by calling <see cref="SetTarget(T)"/>, the previous target may be immediately eligible for garbage collection.
+    /// </remarks>
+    /// <typeparam name="T">The type of object referenced.</typeparam>
+    internal sealed class SoftReference<T>
+        where T : class?
+    {
+        private readonly WeakReference<T?> _weakReference = new(null);
+        private T? _strongReference;
+
+        public SoftReference(T? target)
+        {
+            Gen2GcCallback.Register(
+                static self => ((SoftReference<T>)self).OnGen2GC(),
+                this);
+
+            if (target is not null)
+                SetTarget(target);
```

Comment:
Shouldn't this be done before the Register call?  In theory this thread could be interrupted indefinitely after the Register and before this line, in which case target could actually make its way to Gen2 and OnGen2GC should be able to null out the _strongReference field.  Not sure it really matters.

Reply:
Couldn't that happen essentially anywhere, with the same result?

Reply:
That's why I said "not sure it really matters". But there also doesn't seem to be any reason to do this second,  so why not move it to be first?

Reply:
Now moved